File: BackEnd/model/NoteBook.js
import mongoose from 'mongoose';

const NotebookSchema = new mongoose.Schema({
    user_id: { 
        type: mongoose.Schema.Types.ObjectId, 
        ref: 'User', 
        required: true,
        index: true 
    },
    
    // Ghi chú tự do của user
    title: { 
        type: String, 
        required: true,
        trim: true 
    },
    
    content: { 
        type: String, 
        required: true 
    },
    
    type: { 
        type: String, 
        enum: ['general', 'vocabulary', 'grammar', 'kanji', 'lesson'],
        default: 'general'
    },
    
    tags: [{ 
        type: String, 
        trim: true 
    }],
    
    // Liên kết với item (từ vựng, kanji, ngữ pháp, bài học)
    related_item_id: { 
        type: mongoose.Schema.Types.ObjectId 
    },
    
    related_item_type: { 
        type: String,
        enum: ['vocabulary', 'kanji', 'grammar', 'lesson']
    },
    
    // Giữ lại items cho backward compatibility (nếu cần)
    items: [{
        item_id: { type: mongoose.Schema.Types.ObjectId },
        item_type: { type: String, enum: ['Từ Vựng', 'Kanji', 'Ngữ Pháp'] },
        saved_at: { type: Date, default: Date.now },
        note: String
    }]
    
}, { 
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
});

// Indexes
NotebookSchema.index({ user_id: 1, type: 1 });
NotebookSchema.index({ user_id: 1, createdAt: -1 });
NotebookSchema.index({ tags: 1 });

export default mongoose.model('Notebook', NotebookSchema);

File: FrontEnd/lib/models/notebook.dart
class Notebook {
  final String id;
  final String userId;
  final String title;
  final String content;
  final String type; // 'general', 'vocabulary', 'grammar', 'kanji'
  final String? relatedItemId;
  final String? relatedItemType;
  final List<String> tags;
  final DateTime createdAt;
  final DateTime updatedAt;

  Notebook({
    required this.id,
    required this.userId,
    required this.title,
    required this.content,
    required this.type,
    this.relatedItemId,
    this.relatedItemType,
    this.tags = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Notebook.fromJson(Map<String, dynamic> json) {
    return Notebook(
      id: json['_id'] ?? '',
      userId: json['user_id'] ?? '',
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      type: json['type'] ?? 'general',
      relatedItemId: json['related_item_id'],
      relatedItemType: json['related_item_type'],
      tags: json['tags'] != null 
          ? List<String>.from(json['tags'])
          : [],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'content': content,
      'type': type,
      'related_item_id': relatedItemId,
      'related_item_type': relatedItemType,
      'tags': tags,
    };
  }

  Notebook copyWith({
    String? id,
    String? userId,
    String? title,
    String? content,
    String? type,
    String? relatedItemId,
    String? relatedItemType,
    List<String>? tags,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Notebook(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      title: title ?? this.title,
      content: content ?? this.content,
      type: type ?? this.type,
      relatedItemId: relatedItemId ?? this.relatedItemId,
      relatedItemType: relatedItemType ?? this.relatedItemType,
      tags: tags ?? this.tags,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
